src/lib/utils/alignment_buffer.h

```diff
@@ -0,0 +1,205 @@
+/*
+ * Alignment buffer helper
+ * (C) 2023 Jack Lloyd
+ *     2023 René Meusel - Rohde & Schwarz Cybersecurity
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ */
+
+#ifndef BOTAN_ALIGNMENT_BUFFER_H_
+#define BOTAN_ALIGNMENT_BUFFER_H_
+
+#include <botan/concepts.h>
+#include <botan/internal/stl_util.h>
+
+#include <optional>
+#include <span>
+
+namespace Botan {
+
+constexpr bool DEFER_FINAL_BLOCK = true;
+
+/**
+ * @brief Alignment buffer helper
+ *
+ * Many algorithms have an intrinsic block size in which they consume input
+ * data. When streaming arbitrary data chunks to such algorithms we must store
+ * some data intermittently to honor the algorithm's alignment requirements.
+ *
+ * This helper encapsulates such an alignment buffer. The API of this class
+ * is designed to minimize user errors in the algorithm implementations.
+ * Therefore, it is strongly opinionated on its use case. Don't try to use it
+ * for anything but the described circumstance.
+ *
+ * @tparam T  the container type to use for the alignment buffer
+ */
+template <concepts::contiguous_container T>
+class AlignmentBuffer {
```

Comment:
WDYT about making `defer_final_block` a template parameter? Since I don't think we do (or would ever) have a use case where you would toggle between them at runtime.

Reply:
Similarly, but perhaps more contentiously, if we made the buffer size a template parameter then

a) we might gain some minimal performance improvements due to the the block size being known to be fixed, eg when dividing by the buffer size.
b) (more importantly) we could skip using a container entirely and store the buffer in a simple array, avoiding the heap allocation.

This would be less flexible though, which *might* matter for some future case where we are doing runtime dispatch of the specific algorithm based on hardware support. Eg we want to align to 16 bytes if running regular C++ but 128 bytes if we're doing something with SIMD.

Also it would require exposing the internal block size in the headers of the various algorithms, but these headers are now internal only anyway so this doesn't seem so bad.

Reply:
I'm certainly in favor of making things more "`constexpr`" here. Both the block size and the `defer_final_block` make sense as template parameters. Also I've been tinkering with a template-customizable `append()` function (to allow using this in Keccak where -- among other things -- we'd need to XOR instead of "overwrite"). The latter is certainly future work, though.

Regarding `std::array<>`: I had that in mind, but didn't go for it because we can't use the mlocked allocator for stack allocations. Is that something we can (and want to) live with?

Reply:
I don't think the mlock aspect is that critical tbh. We spray intermediate values everywhere on the stack in the course of computing anything, and it's pretty unavoidable that someone with access to a core dump or swap file is going to be able to recover information.

If we go the array route should in any case scrub the memory on `clear` or destructor; that already puts us ahead of the game compared to most implementations I've seen.

Reply:
Fair enough. I'll look into that tomorrow.

Reply:
:+1:

Reply:
I applied both suggestions, making `defer_final_block` and the `block_size` into template parameters. The latter proved problematic for `MDx_HashFunction` as this info is currently a runtime information (behind a virtual method). Once #3550 is finished, that shouldn't be a problem anymore, though.